nats/nats.go

```diff
@@ -0,0 +1,368 @@
+		)
+	} else {
+		s.logInfow("disconnected",
+			"diver", "nats",
+		)
+	}
+	nc.Opts.RetryOnFailedConnect = true
+	if err != nil {
+		s.err = append(s.err, err)
+	}
+}
+
+// reconnectHandler is a helper function to set the reconnect handler
+func (s *Storage) reconnectHandler(nc *nats.Conn) {
+	s.connectHandler(nc)
+}
+
+// errorHandler is a helper function to set the error handler
+func (s *Storage) errorHandler(nc *nats.Conn, sub *nats.Subscription, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.logErrorw("error handler",
+		"diver", "nats",
+		"sub", sub.Subject,
+		"error", err.Error(),
+	)
+	if err != nil {
+		s.err = append(s.err, fmt.Errorf("subject %q: %w", sub.Subject, err))
+	}
+}
+
+// closedHandler is a helper function to set the closed handler
+func (s *Storage) closedHandler(nc *nats.Conn) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	s.logInfow("closed",
+		"diver", "nats",
+	)
+}
+
+func newNatsKV(nc *nats.Conn, ctx context.Context, keyValueConfig jetstream.KeyValueConfig) (jetstream.KeyValue, error) {
+	js, err := jetstream.New(nc)
+	if err != nil {
+		return nil, fmt.Errorf("get jetstream: %w", err)
+	}
+	jskv, err := js.KeyValue(ctx, keyValueConfig.Bucket)
+	if err != nil {
+		if errors.Is(err, jetstream.ErrBucketNotFound) {
+			jskv, err = js.CreateKeyValue(ctx, keyValueConfig)
+			if err != nil {
+				return nil, fmt.Errorf("jetstream: create kv: %w", err)
+			}
+		} else {
+			return nil, fmt.Errorf("jetstream: get kv: %w", err)
+		}
+	}
+
+	return jskv, nil
+}
+
+// Process the url string argument to Connect.
+// Return an array of urls, even if only one.
+func processUrlString(url string) []string {
+	urls := strings.Split(url, ",")
+	var j int
+	for _, s := range urls {
+		u := strings.TrimSpace(s)
+		if len(u) > 0 {
+			urls[j] = u
+			j++
+		}
+	}
+	return urls[:j]
+}
+
+// New creates a new nats kv storage
+func New(config ...Config) *Storage {
+	// Set default config
+	cfg := configDefault(config...)
+
+	storage := &Storage{
+		ctx: cfg.Context,
+		cfg: cfg,
+	}
+
+	// Set the nats options with default custom handlers
+	cfg.NatsOptions = append(
+		[]nats.Option{
+			nats.ConnectHandler(storage.connectHandler),
+			nats.DisconnectErrHandler(storage.disconnectErrHandler),
+			nats.ReconnectHandler(storage.reconnectHandler),
+			nats.ErrorHandler(storage.errorHandler),
+			nats.ClosedHandler(storage.closedHandler),
+		},
+		cfg.NatsOptions...,
+	)
+	natsOpts := nats.GetDefaultOptions()
+	natsOpts.Servers = processUrlString(cfg.URLs)
+	for _, opt := range cfg.NatsOptions {
+		if opt != nil {
+			if err := opt(&natsOpts); err != nil {
+				panic(err)
+			}
+		}
+	}
+	// Connect to NATS
+	var err error
+	storage.nc, err = natsOpts.Connect()
+
+	if opErr, ok := err.(*net.OpError); ok && natsOpts.RetryOnFailedConnect {
+		if opErr.Op != "dial" {
+			panic(err)
+		}
+	} else if err != nil {
+		panic(err)
+	}
+
+	// TODO improve this crude way to wait for the connection to be established
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Maybe make this a config value. Could be named `ConnectTimeout` with a default val of 100ms

Reply:
It is not technically a ConnectTimeout, just a delay to allow the async handlers to get a connection and create the KV instance.

Reply:
Yup!